modules/configuration/src/main/java/org/apache/ignite/configuration/json/JsonPrimitiveConfigurationSource.java

```diff
@@ -0,0 +1,212 @@
+        return new IllegalArgumentException(String.format(
+            "'%s' is expected as a type for the '%s' configuration value",
+            unbox(clazz).getSimpleName(),
+            join(path) + (idx == -1 ? "" : ("[" + idx + "]"))
+        ));
+    }
+
+    /**
+     * Non-null wrapper over {@link TypeUtils#boxed}.
+     */
+    private static Class<?> box(Class<?> clazz) {
+        Class<?> boxed = TypeUtils.boxed(clazz);
+
+        return boxed == null ? clazz : boxed;
+    }
+
+    /**
+     * Non-null wrapper over {@link TypeUtils#unboxed}.
+     */
+    private static Class<?> unbox(Class<?> clazz) {
+        assert !clazz.isPrimitive();
+
+        Class<?> unboxed = TypeUtils.unboxed(clazz);
+
+        return unboxed == null ? clazz : unboxed;
+    }
+
+    /**
+     * Extracts the value from the given {@link JsonPrimitive} based on the expected type.
+     */
+    private <T> T unwrap(JsonPrimitive jsonPrimitive, Class<T> clazz, int idx) {
+        assert !clazz.isArray();
+        assert !clazz.isPrimitive();
+
+        if (clazz == String.class) {
+            if (!jsonPrimitive.isString())
+                throw wrongTypeException(clazz, idx);
+
+            return clazz.cast(jsonPrimitive.getAsString());
+        }
+        else if (clazz == Boolean.class) {
+            if (!jsonPrimitive.isBoolean())
+                throw wrongTypeException(clazz, idx);
+
+            return clazz.cast(jsonPrimitive.getAsBoolean());
+        }
+        else if (clazz == Character.class) {
+            if (!jsonPrimitive.isString() || jsonPrimitive.getAsString().length() != 1)
+                throw wrongTypeException(clazz, idx);
+
+            return clazz.cast(jsonPrimitive.getAsCharacter());
+        }
+        else if (Number.class.isAssignableFrom(clazz)) {
+            if (!jsonPrimitive.isNumber())
+                throw wrongTypeException(clazz, idx);
+
+            if (clazz == Byte.class) {
+                checkBounds(jsonPrimitive, Byte.MIN_VALUE, Byte.MAX_VALUE);
+
+                return clazz.cast(jsonPrimitive.getAsByte());
+            }
+            else if (clazz == Short.class) {
+                checkBounds(jsonPrimitive, Short.MIN_VALUE, Short.MAX_VALUE);
+
+                return clazz.cast(jsonPrimitive.getAsShort());
+            }
+            else if (clazz == Integer.class) {
+                checkBounds(jsonPrimitive, Integer.MIN_VALUE, Integer.MAX_VALUE);
+
+                return clazz.cast(jsonPrimitive.getAsInt());
+            }
+            else if (clazz == Long.class)
+                return clazz.cast(jsonPrimitive.getAsLong());
+            else if (clazz == Float.class) {
+                double doubleValue = jsonPrimitive.getAsDouble();
+
+                if ((float)doubleValue != doubleValue) {
+                    // don't throw an exception, because double to float conversion is tricky, it is sometimes
+                    // ok to lose precision here
+                    LOG.warn(
```

Comment:
I think we can skip this warning. Long to double conversion loses precision as well but we ignore it, right?

Reply:
maybe we can leave it at the `DEBUG` level? I'm worried that it may lead to some subtle bugs in the future (same applies to the long-to-double conversion)

Reply:
ok, I've given it some thoughts and decided that we can simply point out the possible precision loss in the documentation